src/pipecat/utils/watchdog_async_iterator.py

```diff
@@ -0,0 +1,60 @@
+#
+# Copyright (c) 2024–2025, Daily
+#
+# SPDX-License-Identifier: BSD 2-Clause License
+#
+
+import asyncio
+from typing import AsyncIterator, Optional
+
+from pipecat.utils.watchdog_reseter import WatchdogReseter
+
+
+class WatchdogAsyncIterator:
+    """An asynchronous iterator that monitors activity and resets the current
+    task watchdog timer. This is necessary to avoid task watchdog timers to
+    expire while we are waiting to get an item from the iterator.
+
+    """
+
+    def __init__(self, async_iterable, *, reseter: WatchdogReseter, timeout: float = 2.0):
+        self._async_iterable = async_iterable
+        self._reseter = reseter
+        self._timeout = timeout
+        self._iter: Optional[AsyncIterator] = None
+        self._current_anext_task: Optional[asyncio.Task] = None
+
+    def __aiter__(self):
+        return self
+
+    async def __anext__(self):
+        if not self._iter:
+            self._iter = await self._ensure_async_iterator(self._async_iterable)
+
+        while True:
+            try:
+                if not self._current_anext_task:
+                    self._current_anext_task = asyncio.create_task(self._iter.__anext__())
+
+                item = await asyncio.wait_for(
+                    asyncio.shield(self._current_anext_task),
+                    timeout=self._timeout,
+                )
+
+                self._reseter.reset_watchdog()
+
+                # The task has finish, so we will create a new one for th next item.
+                self._current_anext_task = None
+
+                return item
+            except asyncio.TimeoutError:
+                self._reseter.reset_watchdog()
+            except StopAsyncIteration:
+                self._current_anext_task = None
+                raise
```

Comment:
If we change it like this, we are going to prevent the exception that we are receiving when ending the pipeline:

```suggestion
            except StopAsyncIteration:
                # Await or retrieve the task result to suppress asyncio warning
                if self._current_anext_task and self._current_anext_task.done():
                    try:
                        self._current_anext_task.result()
                    except StopAsyncIteration:
                        pass
                self._current_anext_task = None
                raise
```